# Welcome to Assignment 2

In this assignment, we would love you to explore the dataset c.elegans, a dataset of C. Elegans' neural network. It would be an excellent opportunity to get familiar with Networkx. 

In [1]:
import networkx as nx
import pandas as pd

First, let's start reading the data. Remember that every time you read network data, make sure the type of networks you want to create. The functions provided for networks are different depending on their types. In this assignment, the one we will explore is directed weighted network, `class nx.Digraph`.

There is the canvas link (https://iu.instructure.com/files/114453534/download?download_frd=1) for downloading c.elegans. Or you can access the data through the FirstCourseNetworkScience (https://github.com/CambridgeUniversityPress/FirstCourseNetworkScience) git repo. 


Let's take a look at the data first before we decide which method for reading data. 

In [2]:
# make sure to repalce the path of the data when you run the code
edges = pd.read_csv('..\\FirstCourseNetworkScience\\datasets\\celegansneural\\celegansneural.edges', sep = ' ')

In [3]:
edges.head()

#  Directed,  weighted
0  135          1         3
1  202          1         2
2    1          2         1
3    8          2         1
4   17          2         6

In [4]:
len(edges)

2345

Now we see the structure of edge data, where the first two columns refer to nodes and the third column refers to the weight of the corresponding edge. It should be noted that the first column stores the outcoming nodes, while the second column stores the incoming nodes. 

In [5]:
G = nx.read_weighted_edgelist('..\\FirstCourseNetworkScience\\datasets\\celegansneural\\celegansneural.edges',
                             create_using = nx.DiGraph())

It should be noted that the type of node name, in this case, is a string instead of int.

In [6]:
G.has_edge(135,1)

False

In [7]:
G.has_edge('135', '1')

True

In [8]:
G.has_edge('1', '135')

False

Now you have the network object, let's start exploring it a little bit. Please answer the following questions. Interpretation of the result would be appreciated. :)

## Q1. How many nodes and links are in the network G?

In [9]:
G.number_of_nodes()

297

In [10]:
G.number_of_edges()

2345

There are 2345 edges as specified in the input file. As there could be more than one outgoing or incoming links from a node the number of nodes(297) is < number of edges(2345).

## Q2. What is the density of network G?

In [11]:
# Density is the ratio of actual nummber of links to maximum possible number of links.
L = G.number_of_edges()
N = G.number_of_nodes()
d = (L/(N*(N - 1))) # density
print(d)

0.026674401674401674


As the density is less than 1 it means it is a sparse network i.e. not all nodes are connected to all other nodes.

## Q3. Please show the node with the largest in-degree, and the node with the largest out-degree.

In [21]:
print("Node with largest Degree:", max(G.degree()))

Node with largest Degree: ('99', 34)


In [22]:
print("Node with largest IN Degree:",max(G.in_degree()))

Node with largest IN Degree: ('99', 19)


In [23]:
print("Node with largest OUT Degree:",max(G.out_degree()))

Node with largest OUT Degree: ('99', 15)


## Q4. Please show the node with the largest in-strength, and the one with the largest out-strength.

## Q5. Please calculate the average in-degree and out-degree of the network G. Briefly interpret your finding.

## Q6. Draw the network.

## Q7. Are there any nodes with zero in-degree? How many?

## Q8. Are there any nodes with zero out-degree? How many?

There comes the fun parts. Time to explore common neighbours. 

## Q9. Which pair of nodes have the largest number of common predecessors?

## Q10. Which pair of nodes have the largest number of common successors?

## For the following questions, please read c.elegans as undirected network.

You can reread the c.elegans file, or try nx.Graph(G) to force the directed graph into an undirected graph.

## Q11. What's the average path length?

## Q12. What's the diameter?

## Q13. Which node has the largest clustering coefficient?

## Q14. What's the average clustering coefficient?